<a href="https://cognitiveclass.ai"><img src = "https://ibm.box.com/shared/static/9gegpsmnsoo25ikkbl4qzlvlyjbgxs5x.png" width = 400> </a>

<h1 align=center><font size = 5>From Understanding to Preparation</font></h1>

## Introduction

In this lab, we will continue learning about the data science methodology, and focus on the **Data Understanding** and the **Data Preparation** stages.

## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">
1. [Recap](#0)<br>
2. [Data Understanding](#2)<br>
3. [Data Preparation](#4)<br>
</div>
<hr>

# Recap(Summary) <a id="0"></a>

In Lab **From Requirements to Collection**, we learned that the data we need to answer the question developed in the business understanding stage, namely *can we automate the process of determining the cuisine of a given recipe?*, is readily available. A researcher named Yong-Yeol Ahn scraped tens of thousands of food recipes (cuisines and ingredients) from three different websites, namely:

<img src = "https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DS0103EN/labs/images/lab3_fig1_allrecipes.png" width=500>

www.allrecipes.com

<img src = "https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DS0103EN/labs/images/lab3_fig2_epicurious.png" width=500>

www.epicurious.com

<img src = "https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DS0103EN/labs/images/lab3_fig3_menupan.png" width=500>

www.menupan.com

For more information on Yong-Yeol Ahn and his research, you can read his paper on [Flavor Network and the Principles of Food Pairing](http://yongyeol.com/papers/ahn-flavornet-2011.pdf).

We also collected the data and placed it on an IBM server for your convenience.

------------

# Data Understanding <a id="2"></a>

<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DS0103EN/labs/images/lab3_fig4_flowchart_data_understanding.png" width=500>

<strong> Important note:</strong> Please note that you are not expected to know how to program in Python. The following code is meant to illustrate the stages of data understanding and data preparation, so it is totally fine if you do not understand the individual lines of code. We have a full course on programming in Python, <a href="http://cocl.us/PY0101EN_DS0103EN_LAB3_PYTHON_Coursera"><strong>Python for Data Science</strong></a>, which is also offered on Coursera. So make sure to complete the Python course if you are interested in learning how to program in Python.

### Using this notebook:

To run any of the following cells of code, you can type **Shift + Enter** to excute the code in a cell.

Get the version of Python installed.

In [1]:
# check Python version
!python -V

Python 3.7.4


Download the library and dependencies that we will need to run this lab.

In [2]:
import pandas as pd # import library to read data into dataframe
pd.set_option('display.max_columns', None)
import numpy as np # import numpy library
import re # import library for regular expression

Download the data from the IBM server and read it into a *pandas* dataframe.

In [3]:
recipes = pd.read_csv("https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DS0103EN/labs/data/recipes.csv")

print("Data read into dataframe!") # takes about 30 seconds

Data read into dataframe!


Show the first few rows.

In [28]:
recipes.head()

0              0   
2       0       0              0        0        0              0   
3       0       0              0        0        0              0   
4       0       0              0        0        0              0   

   chicken_liver  chickpea  chicory  chinese_cabbage  chive  cider  cilantro  \
0              0         0        0                0      0      0         1   
1              0         0        0                0      0      0         0   
2              0         0        0                0      0      0         0   
3              0         0        0                0      0      0         1   
4              0         0        0                0      0      0         0   

   cinnamon  citrus  citrus_peel  clam  clove  cocoa  coconut  coconut_oil  \
0         0       0            0     0      0      0        0            0   
1         0       0            0     0      0      0        0            0   
2         0       0            0     0      0      0        0            0   
3         0       0            0     0      0      0        0            0   
4         0       0            0     0      0      0        0            0   

   cod  coffee  cognac  concord_grape  condiment  coriander  corn  corn_flake  \
0    0       0       0              0          0          0     0           0   
1    0       0       0              0          0          0     0           0   
2    0       0       0              0          0          0     0           0   
3    0       0       0              0          0          0     0           0   
4    0       0       0              0          0          1     0           0   

   corn_grit  cottage_cheese  crab  cranberry  cream  cream_cheese  cucumber  \
0          0               0     0          0      0             0         1   
1          0               0     0          0      0             0         0   
2          0               0     0          0      0             0         0   
3          0               0     0          0      0             0         0   
4          0               0     0          0      0             0         1   

   cumin  cured_pork  currant  date  dill  durian  eel  egg  egg_noodle  \
0      0           0        0     0     0       0    0    0           0   
1      0           0        0     0     0       0    0    0           0   
2      0           0        0     0     0       0    0    0           0   
3      0           0        0     0     0       0    0    0           0   
4      0           0        0     0     0       0    0    0           0   

   elderberry  emmental_cheese  endive  enokidake  fennel  fenugreek  \
0           0                0       0          0       0          0   
1           0                0       0          0       0          0   
2           0                0       0          0       0          0   
3           0                0       0          0       0          0   
4           0                0       0          0       0          0   

   feta_cheese  fig  fish  flower  frankfurter  fruit  galanga  gardenia  \
0            0    0     1       0            0      0        0         0   
1            0    0     1       0            0      0        0         0   
2            0    0     0       0            0      0        0         0   
3            0    0     1       0            0      0        0         0   
4            0    0     1       0            0      0        0         0   

   garlic  gelatin  geranium  gin  ginger  goat_cheese  grape  grape_brandy  \
0       1        0         0    0       0            0      0             0   
1       1        0         0    0       0            0      0             0   
2       1        0         0    0       0            0      0             0   
3       0        0         0    0       1            0      0             0   
4       1        0         0    0       0            0      0             0   

   grape_juice  grapefruit  green_bell

Get the dimensions of the dataframe.

In [5]:
recipes.shape

(57691, 384)

So our dataset consists of 57,691 recipes. Each row represents a recipe, and for each recipe, the corresponding cuisine is documented as well as whether 384 ingredients exist in the recipe or not, beginning with almond and ending with zucchini.

We know that a basic sushi recipe includes the ingredients:
* rice
* soy sauce
* wasabi
* some fish/vegetables

Let's check that these ingredients exist in our dataframe:

In [6]:
ingredients = list(recipes.columns.values)

print([match.group(0) for ingredient in ingredients for match in [(re.compile(".*(rice).*")).search(ingredient)] if match])
print([match.group(0) for ingredient in ingredients for match in [(re.compile(".*(wasabi).*")).search(ingredient)] if match])
print([match.group(0) for ingredient in ingredients for match in [(re.compile(".*(soy).*")).search(ingredient)] if match])

['brown_rice', 'licorice', 'rice']
['wasabi']
['soy_sauce', 'soybean', 'soybean_oil']


Yes, they do!

* rice exists as rice.
* wasabi exists as wasabi.
* soy exists as soy_sauce.

So maybe if a recipe contains all three ingredients: rice, wasabi, and soy_sauce, then we can confidently say that the recipe is a **Japanese** cuisine! Let's keep this in mind!

----------------

# Data Preparation <a id="4"></a>

<img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DS0103EN/labs/images/lab3_fig5_flowchart_data_preparation.png" width=500>

In this section, we will prepare data for the next stage in the data science methodology, which is modeling. This stage involves exploring the data further and making sure that it is in the right format for the machine learning algorithm that we selected in the analytic approach stage, which is decision trees.

First, look at the data to see if it needs cleaning.

In [7]:
recipes["country"].value_counts() # frequency table

American        40150
Mexico           1754
Italian          1715
Italy            1461
Asian            1176
                ...  
Indonesia          12
Belgium            11
East-African       11
Israel              9
Bangladesh          4
Name: country, Length: 69, dtype: int64

By looking at the above table, we can make the following observations:

1. Cuisine column is labeled as Country, which is inaccurate.
2. Cuisine names are not consistent as not all of them start with an uppercase first letter.
3. Some cuisines are duplicated as variation of the country name, such as Vietnam and Vietnamese.
4. Some cuisines have very few recipes.

#### Let's fixes these problems.

Fix the name of the column showing the cuisine.

In [8]:
column_names = recipes.columns.values
column_names[0] = "cuisine"
recipes.columns = column_names

recipes

57688  No      No      No   No    No        No    No     No            No   
57689  No      No      No   No    No        No    No     No            No   
57690  No      No      No   No    No        No   Yes     No            No   

      orange_juice orange_peel oregano ouzo oyster palm papaya  \
0               No          No      No   No     No   No     No   
1               No          No      No   No     No   No     No   
2               No          No      No   No     No   No     No   
3               No          No      No   No     No   No     No   
4               No          No      No   No     No   No     No   
...            ...         ...     ...  ...    ...  ...    ...   
57686           No          No      No   No     No   No     No   
57687           No          No      No   No     No   No     No   
57688           No          No      No   No     No   No     No   
57689           No          No      No   No     No   No     No   
57690           No          No      No   No     No   No     No   

      parmesan_cheese parsley parsnip passion_fruit  pea peach peanut  \
0                  No      No      No            No   No    No     No   
1                  No      No      No            No   No    No     No   
2                  No      No      No            No   No    No     No   
3                  No      No      No            No  Yes    No     No   
4                  No      No      No            No   No    No    Yes   
...               ...     ...     ...           ...  ...   ...    ...   
57686              No     Yes      No            No   No    No     No   
57687              No      No      No            No   No    No     No   
57688              No      No      No            No   No    No     No   
57689              No      No      No            No   No    No     No   
57690              No      No      No            No   No    No     No   

      peanut_butter peanut_oil pear pear_brandy pecan pelargonium pepper  \
0                No         No   No          No    No          No     No   
1                No         No   No          No    No          No     No   
2                No         No   No          No    No          No     No   
3                No         No   No          No    No          No     No   
4                No         No   No          No    No          No     No   
...             ...        ...  ...         ...   ...         ...    ...   
57686            No         No   No          No    No          No     No   
57687            No         No   No          No    No          No     No   
57688            No         No   No          No    No          No    Yes   
57689            No         No   No          No    No          No     No   
57690            No         No   No          No    No          No    Yes   

      peppermint peppermint_oil pimenta pimento pineapple pistachio plum  \
0             No             No      No      No        No        No   No   
1             No             No      No      No        No        No   No   
2             No             No      No      No        No        No   No   
3             No             No      No      No        No        No   No   
4             No             No      No      No        No        No   No   
...          ...            ...     ...     ...       ...       ...  ...   
57686         No             No      No      No        No        No   No   
57687         No             No      No      No        No        No   No   
57688         No             No      No      No        No        No   No   
57689         No             No      No      No        No        No   No   
57690         No             No      No      No        No        No   No   

      popcorn porcini pork pork_liver pork_sausage port_wine potato  \
0          No      No   No         No           No        No     No   
1          No      No   No         No           No        No     No   
2          No      No   No         No           No        N

Make all the cuisine names lowercase.

In [9]:
recipes["cuisine"] = recipes["cuisine"].str.lower()

Make the cuisine names consistent.

In [10]:
recipes.loc[recipes["cuisine"] == "austria", "cuisine"] = "austrian"
recipes.loc[recipes["cuisine"] == "belgium", "cuisine"] = "belgian"
recipes.loc[recipes["cuisine"] == "china", "cuisine"] = "chinese"
recipes.loc[recipes["cuisine"] == "canada", "cuisine"] = "canadian"
recipes.loc[recipes["cuisine"] == "netherlands", "cuisine"] = "dutch"
recipes.loc[recipes["cuisine"] == "france", "cuisine"] = "french"
recipes.loc[recipes["cuisine"] == "germany", "cuisine"] = "german"
recipes.loc[recipes["cuisine"] == "india", "cuisine"] = "indian"
recipes.loc[recipes["cuisine"] == "indonesia", "cuisine"] = "indonesian"
recipes.loc[recipes["cuisine"] == "iran", "cuisine"] = "iranian"
recipes.loc[recipes["cuisine"] == "italy", "cuisine"] = "italian"
recipes.loc[recipes["cuisine"] == "japan", "cuisine"] = "japanese"
recipes.loc[recipes["cuisine"] == "israel", "cuisine"] = "jewish"
recipes.loc[recipes["cuisine"] == "korea", "cuisine"] = "korean"
recipes.loc[recipes["cuisine"] == "lebanon", "cuisine"] = "lebanese"
recipes.loc[recipes["cuisine"] == "malaysia", "cuisine"] = "malaysian"
recipes.loc[recipes["cuisine"] == "mexico", "cuisine"] = "mexican"
recipes.loc[recipes["cuisine"] == "pakistan", "cuisine"] = "pakistani"
recipes.loc[recipes["cuisine"] == "philippines", "cuisine"] = "philippine"
recipes.loc[recipes["cuisine"] == "scandinavia", "cuisine"] = "scandinavian"
recipes.loc[recipes["cuisine"] == "spain", "cuisine"] = "spanish_portuguese"
recipes.loc[recipes["cuisine"] == "portugal", "cuisine"] = "spanish_portuguese"
recipes.loc[recipes["cuisine"] == "switzerland", "cuisine"] = "swiss"
recipes.loc[recipes["cuisine"] == "thailand", "cuisine"] = "thai"
recipes.loc[recipes["cuisine"] == "turkey", "cuisine"] = "turkish"
recipes.loc[recipes["cuisine"] == "vietnam", "cuisine"] = "vietnamese"
recipes.loc[recipes["cuisine"] == "uk-and-ireland", "cuisine"] = "uk-and-irish"
recipes.loc[recipes["cuisine"] == "irish", "cuisine"] = "uk-and-irish"

recipes

57688  No      No      No   No    No        No    No     No            No   
57689  No      No      No   No    No        No    No     No            No   
57690  No      No      No   No    No        No   Yes     No            No   

      orange_juice orange_peel oregano ouzo oyster palm papaya  \
0               No          No      No   No     No   No     No   
1               No          No      No   No     No   No     No   
2               No          No      No   No     No   No     No   
3               No          No      No   No     No   No     No   
4               No          No      No   No     No   No     No   
...            ...         ...     ...  ...    ...  ...    ...   
57686           No          No      No   No     No   No     No   
57687           No          No      No   No     No   No     No   
57688           No          No      No   No     No   No     No   
57689           No          No      No   No     No   No     No   
57690           No          No      No   No     No   No     No   

      parmesan_cheese parsley parsnip passion_fruit  pea peach peanut  \
0                  No      No      No            No   No    No     No   
1                  No      No      No            No   No    No     No   
2                  No      No      No            No   No    No     No   
3                  No      No      No            No  Yes    No     No   
4                  No      No      No            No   No    No    Yes   
...               ...     ...     ...           ...  ...   ...    ...   
57686              No     Yes      No            No   No    No     No   
57687              No      No      No            No   No    No     No   
57688              No      No      No            No   No    No     No   
57689              No      No      No            No   No    No     No   
57690              No      No      No            No   No    No     No   

      peanut_butter peanut_oil pear pear_brandy pecan pelargonium pepper  \
0                No         No   No          No    No          No     No   
1                No         No   No          No    No          No     No   
2                No         No   No          No    No          No     No   
3                No         No   No          No    No          No     No   
4                No         No   No          No    No          No     No   
...             ...        ...  ...         ...   ...         ...    ...   
57686            No         No   No          No    No          No     No   
57687            No         No   No          No    No          No     No   
57688            No         No   No          No    No          No    Yes   
57689            No         No   No          No    No          No     No   
57690            No         No   No          No    No          No    Yes   

      peppermint peppermint_oil pimenta pimento pineapple pistachio plum  \
0             No             No      No      No        No        No   No   
1             No             No      No      No        No        No   No   
2             No             No      No      No        No        No   No   
3             No             No      No      No        No        No   No   
4             No             No      No      No        No        No   No   
...          ...            ...     ...     ...       ...       ...  ...   
57686         No             No      No      No        No        No   No   
57687         No             No      No      No        No        No   No   
57688         No             No      No      No        No        No   No   
57689         No             No      No      No        No        No   No   
57690         No             No      No      No        No        No   No   

      popcorn porcini pork pork_liver pork_sausage port_wine potato  \
0          No      No   No         No           No        No     No   
1          No      No   No         No           No        No     No   
2          No      No   No         No           No        N

Remove cuisines with < 50 recipes.

In [11]:
# get list of cuisines to keep
recipes_counts = recipes["cuisine"].value_counts()
cuisines_indices = recipes_counts > 50

cuisines_to_keep = list(np.array(recipes_counts.index.values)[np.array(cuisines_indices)])

In [12]:
rows_before = recipes.shape[0] # number of rows of original dataframe
print("Number of rows of original dataframe is {}.".format(rows_before))

recipes = recipes.loc[recipes['cuisine'].isin(cuisines_to_keep)]

rows_after = recipes.shape[0] # number of rows of processed dataframe
print("Number of rows of processed dataframe is {}.".format(rows_after))

print("{} rows removed!".format(rows_before - rows_after))

Number of rows of original dataframe is 57691.
Number of rows of processed dataframe is 57403.
288 rows removed!


Convert all Yes's to 1's and the No's to 0's

In [13]:
recipes = recipes.replace(to_replace="Yes", value=1)
recipes = recipes.replace(to_replace="No", value=0)

#### Let's analyze the data a little more in order to learn the data better and note any interesting preliminary observations.

Run the following cell to get the recipes that contain **rice** *and* **soy** *and* **wasabi** *and* **seaweed**.

In [14]:
recipes.head()

0              0   
2       0       0              0        0        0              0   
3       0       0              0        0        0              0   
4       0       0              0        0        0              0   

   chicken_liver  chickpea  chicory  chinese_cabbage  chive  cider  cilantro  \
0              0         0        0                0      0      0         1   
1              0         0        0                0      0      0         0   
2              0         0        0                0      0      0         0   
3              0         0        0                0      0      0         1   
4              0         0        0                0      0      0         0   

   cinnamon  citrus  citrus_peel  clam  clove  cocoa  coconut  coconut_oil  \
0         0       0            0     0      0      0        0            0   
1         0       0            0     0      0      0        0            0   
2         0       0            0     0      0      0        0            0   
3         0       0            0     0      0      0        0            0   
4         0       0            0     0      0      0        0            0   

   cod  coffee  cognac  concord_grape  condiment  coriander  corn  corn_flake  \
0    0       0       0              0          0          0     0           0   
1    0       0       0              0          0          0     0           0   
2    0       0       0              0          0          0     0           0   
3    0       0       0              0          0          0     0           0   
4    0       0       0              0          0          1     0           0   

   corn_grit  cottage_cheese  crab  cranberry  cream  cream_cheese  cucumber  \
0          0               0     0          0      0             0         1   
1          0               0     0          0      0             0         0   
2          0               0     0          0      0             0         0   
3          0               0     0          0      0             0         0   
4          0               0     0          0      0             0         1   

   cumin  cured_pork  currant  date  dill  durian  eel  egg  egg_noodle  \
0      0           0        0     0     0       0    0    0           0   
1      0           0        0     0     0       0    0    0           0   
2      0           0        0     0     0       0    0    0           0   
3      0           0        0     0     0       0    0    0           0   
4      0           0        0     0     0       0    0    0           0   

   elderberry  emmental_cheese  endive  enokidake  fennel  fenugreek  \
0           0                0       0          0       0          0   
1           0                0       0          0       0          0   
2           0                0       0          0       0          0   
3           0                0       0          0       0          0   
4           0                0       0          0       0          0   

   feta_cheese  fig  fish  flower  frankfurter  fruit  galanga  gardenia  \
0            0    0     1       0            0      0        0         0   
1            0    0     1       0            0      0        0         0   
2            0    0     0       0            0      0        0         0   
3            0    0     1       0            0      0        0         0   
4            0    0     1       0            0      0        0         0   

   garlic  gelatin  geranium  gin  ginger  goat_cheese  grape  grape_brandy  \
0       1        0         0    0       0            0      0             0   
1       1        0         0    0       0            0      0             0   
2       1        0         0    0       0            0      0             0   
3       0        0         0    0       1            0      0             0   
4       1        0         0    0       0            0      0             0   

   grape_juice  grapefruit  green_bell

In [15]:
check_recipes = recipes.loc[
    (recipes["rice"] == 1) &
    (recipes["soy_sauce"] == 1) &
    (recipes["wasabi"] == 1) &
    (recipes["seaweed"] == 1)
]

check_recipes

0     0   
13010              0    0      0       0              0           0     0   
13159              0    0      0       0              0           0     0   
13513              0    0      0       0              0           0     0   
13586              0    0      0       0              0           0     0   
13625              0    0      0       0              0           0     0   
14495              0    0      0       0              0           0     0   

       pear_brandy  pecan  pelargonium  pepper  peppermint  peppermint_oil  \
11306            0      0            0       0           0               0   
11321            0      0            0       1           0               0   
11361            0      0            0       0           0               0   
12171            0      0            0       1           0               0   
12385            0      0            0       0           0               0   
13010            0      0            0       0           0               0   
13159            0      0            0       0           0               0   
13513            0      0            0       0           0               0   
13586            0      0            0       0           0               0   
13625            0      0            0       0           0               0   
14495            0      0            0       0           0               0   

       pimenta  pimento  pineapple  pistachio  plum  popcorn  porcini  pork  \
11306        0        0          0          0     0        0        0     0   
11321        0        0          0          0     0        0        0     0   
11361        0        0          0          0     0        0        0     0   
12171        0        0          0          0     0        0        0     0   
12385        0        0          0          0     0        0        0     0   
13010        0        0          0          0     0        0        0     0   
13159        0        0          0          0     0        0        0     0   
13513        0        0          0          0     0        0        0     0   
13586        0        0          0          0     0        0        0     0   
13625        0        0          0          0     0        0        0     0   
14495        0        0          0          0     0        0        0     0   

       pork_liver  pork_sausage  port_wine  potato  potato_chip  prawn  \
11306           0             0          0       0            0      0   
11321           0             0          0       0            0      0   
11361           0             0          0       0            0      0   
12171           0             0          0       0            0      0   
12385           0             0          0       0            0      0   
13010           0             0          0       0            0      0   
13159           0             0          0       0            0      0   
13513           0             0          0       0            0      0   
13586           0             0          0       0            0      0   
13625           0             0          0       0            0      0   
14495           0             0          0       0            0      0   

       prickly_pear  provolone_cheese  pumpkin  quince  radish  raisin  \
11306             0                 0        0       0       0       0   
11321             0                 0        0       0       0       0   
11361             0                 0        0       0       0       0   
12171             0                 0        0       0       0       0   
12385             0                 0        0       0       0       0   
13010             0                 0        0       0       0       0   
13159             0                 0        0       0       0       0   
13513             0                 0        0       0       1       0   
13586             0                 0        0       0       1       0   
13625      

Based on the results of the above code, can we classify all recipes that contain **rice** *and* **soy** *and* **wasabi** *and* **seaweed** as **Japanese** recipes? Why?

Double-click __here__ for the solution.
<!-- The correct answer is:
No, because other recipes such as Asian and East_Asian recipes also contain these ingredients.
-->

Let's count the ingredients across all recipes.

In [16]:
# sum each column
ing = recipes.iloc[:, 1:].sum(axis=0)

In [17]:
# define each column as a pandas series
ingredient = pd.Series(ing.index.values, index = np.arange(len(ing)))
count = pd.Series(list(ing), index = np.arange(len(ing)))

# create the dataframe
ing_df = pd.DataFrame(dict(ingredient = ingredient, count = count))
ing_df = ing_df[["ingredient", "count"]]
print(ing_df.to_string())

ingredient  count
0                     almond   2306
1                   angelica      1
2                      anise    223
3                 anise_seed     87
4                      apple   2422
5               apple_brandy     37
6                    apricot    620
7                   armagnac     11
8                  artemisia     13
9                  artichoke    391
10                 asparagus    460
11                   avocado    660
12                     bacon   2169
13              baked_potato      9
14                      balm      3
15                    banana    989
16                    barley    266
17             bartlett_pear     23
18                     basil   3842
19                       bay   1463
20                      bean   1992
21                     beech      1
22                      beef   4902
23                beef_broth    845
24                beef_liver     10
25                      beer    307
26                      beet    233
27        

Now we have a dataframe of ingredients and their total counts across all recipes. Let's sort this dataframe in descending order.

In [18]:
ing_df.sort_values(["count"], ascending=False, inplace=True)
ing_df.reset_index(inplace=True, drop=True)

print(ing_df)

ingredient  count
0                egg  21025
1              wheat  20781
2             butter  20719
3              onion  18080
4             garlic  17353
..               ...    ...
378   strawberry_jam      1
379  sturgeon_caviar      1
380      kaffir_lime      1
381            beech      1
382           durian      0

[383 rows x 2 columns]


#### What are the 3 most popular ingredients?

Double-click __here__ for the solution.
<!-- The correct answer is:
// 1. Egg with 21,025 occurrences. 
// 2. Wheat with 20,781 occurrences. 
// 3. Butter with 20,719 occurrences.
-->

However, note that there is a problem with the above table. There are ~40,000 American recipes in our dataset, which means that the data is biased towards American ingredients.

**Therefore**, let's compute a more objective summary of the ingredients by looking at the ingredients per cuisine.

#### Let's create a *profile* for each cuisine.

In other words, let's try to find out what ingredients Chinese people typically use, and what is **Canadian** food for example.

In [19]:
cuisines = recipes.groupby("cuisine").mean()
cuisines.head()

african            0.017391     0.000000  0.000000  0.000000  0.017391   
american           0.001619     0.000075  0.000174  0.000523  0.002117   
asian              0.006706     0.000000  0.002515  0.000838  0.004191   
cajun_creole       0.000000     0.000000  0.000000  0.006849  0.000000   
canadian           0.000000     0.000000  0.000000  0.001292  0.006460   

              long_pepper    lovage  macadamia_nut  macaroni      mace  \
cuisine                                                                  
african          0.000000  0.000000       0.000000  0.026087  0.008696   
american         0.000000  0.003064       0.002466  0.046252  0.002192   
asian            0.001676  0.000000       0.000838  0.036044  0.002515   
cajun_creole     0.000000  0.027397       0.000000  0.020548  0.000000   
canadian         0.000000  0.002584       0.000000  0.054264  0.001292   

              mackerel      malt  mandarin  mandarin_peel     mango  \
cuisine                                                               
african            0.0  0.000000  0.000000       0.000000  0.034783   
american           0.0  0.000672  0.006575       0.000050  0.006002   
asian              0.0  0.000838  0.000838       0.004191  0.026823   
cajun_creole       0.0  0.000000  0.000000       0.000000  0.006849   
canadian           0.0  0.001292  0.005168       0.000000  0.006460   

              maple_syrup  marjoram      mate  matsutake      meat     melon  \
cuisine                                                                        
african          0.000000  0.034783  0.000000   0.000000  0.017391  0.000000   
american         0.010760  0.010610  0.000025   0.000000  0.015866  0.003238   
asian            0.005868  0.000000  0.000000   0.000838  0.017603  0.005029   
cajun_creole     0.000000  0.006849  0.000000   0.000000  0.020548  0.000000   
canadian         0.021964  0.010336  0.000000   0.000000  0.009044  0.003876   

                  milk  milk_fat      mint  mozzarella_cheese  mung_bean  \
cuisine                                                                    
african       0.086957  0.034783  0.060870           0.000000   0.000000   
american      0.266002  0.012603  0.011880           0.018431   0.000025   
asian         0.025147  0.012573  0.082146           0.000838   0.004191   
cajun_creole  0.082192  0.041096  0.006849           0.006849   0.000000   
canadian      0.264858  0.010336  0.006460           0.027132   0.000000   

              munster_cheese  muscat_grape  mushroom    mussel   mustard  \
cuisine                                                                    
african             0.000000      0.000000  0.000000  0.000000  0.052174   
american            0.000598      0.000025  0.060747  0.001046  0.082889   
asian               0.000000      0.000000  0.051970  0.005029  0.024308   
cajun_creole        0.000000      0.000000  0.041096  0.013699  0.143836   
canadian            0.000000      0.000000  0.069767  0.000000  0.085271   

               mutton  nectarine      nira       nut    nutmeg       oat  \
cuisine                                                                    
african       0.00000   0.000000  0.000000  0.026087  0.034783  0.000000   
american      0.00005   0.000996  0.000000  0.018705  0.051731  0.029639   
asian         0.00000   0.000838  0.000838  0.033529  0.008382  0.000000   
cajun_creole  0.00000   0.006849  0.000000  0.006849  0.027397  0.000000   
canadian      0.00000   0.002584  0.000000  0.010336  0.033592  0.033592   

               oatmeal   octopus      okra     olive  olive_oil     onion  \
cuisine                                                                     
african       0.000000  0.008696  0.017391  0.104348   0.521739  0.530435   
american      0.001295  0.000000  0.001569  0.025056   0.132628  0.293325   
asian         0.000838  0.000000  0.001676  0.001676   0.066220  0.145851   
cajun_creole  0.000000  0.000000  0.082192  0.000000   0.287671  0.

As shown above, we have just created a dataframe where each row is a cuisine and each column (except for the first column) is an ingredient, and the row values represent the percentage of each ingredient in the corresponding cuisine.

**For example**:

* *almond* is present across 15.65% of all of the **African** recipes.
* *butter* is present across 38.11% of all of the **Canadian** recipes.

Let's print out the profile for each cuisine by displaying the top four ingredients in each cuisine.

In [20]:
num_ingredients = 4 # define number of top ingredients to print

# define a function that prints the top ingredients for each cuisine
def print_top_ingredients(row):
    print(row.name.upper())
    row_sorted = row.sort_values(ascending=False)*100
    top_ingredients = list(row_sorted.index.values)[0:num_ingredients]
    row_sorted = list(row_sorted)[0:num_ingredients]

    for ind, ingredient in enumerate(top_ingredients):
        print("%s (%d%%)" % (ingredient, row_sorted[ind]), end=' ')
    print("\n")

# apply function to cuisines dataframe
create_cuisines_profiles = cuisines.apply(print_top_ingredients, axis=1)

AFRICAN
onion (53%) olive_oil (52%) garlic (49%) cumin (42%) 

AMERICAN
butter (41%) egg (40%) wheat (39%) onion (29%) 

ASIAN
soy_sauce (49%) ginger (48%) garlic (47%) rice (41%) 

CAJUN_CREOLE
onion (69%) cayenne (56%) garlic (48%) butter (36%) 

CANADIAN
wheat (39%) butter (38%) egg (35%) onion (34%) 

CARIBBEAN
onion (51%) garlic (50%) vegetable_oil (31%) black_pepper (31%) 

CENTRAL_SOUTHAMERICAN
garlic (56%) onion (54%) cayenne (51%) tomato (41%) 

CHINESE
soy_sauce (68%) ginger (53%) garlic (52%) scallion (48%) 

EAST_ASIAN
garlic (55%) soy_sauce (50%) scallion (49%) cayenne (47%) 

EASTERN-EUROPE
wheat (53%) egg (52%) butter (48%) onion (45%) 

EASTERNEUROPEAN_RUSSIAN
butter (60%) egg (50%) wheat (49%) onion (38%) 

ENGLISH_SCOTTISH
butter (67%) wheat (62%) egg (53%) cream (41%) 

FRENCH
butter (50%) egg (44%) wheat (37%) olive_oil (27%) 

GERMAN
wheat (64%) egg (60%) butter (47%) onion (34%) 

GREEK
olive_oil (76%) garlic (44%) onion (36%) lemon_juice (33%) 

INDIAN
cumin (60%

At this point, we feel that we have understood the data well and the data is ready and is in the right format for modeling!

-----------

### Thank you for completing this lab!

This notebook was created by [Alex Aklson](https://www.linkedin.com/in/aklson/). We hope you found this lab session interesting. Feel free to contact us if you have any questions!

This notebook is part of a course on **Coursera** called *Data Science Methodology*. If you accessed this notebook outside the course, you can take this course, online by clicking [here](https://cocl.us/DS0103EN_Coursera_LAB3).

<hr>

Copyright &copy; 2019 [Cognitive Class](https://cognitiveclass.ai/?utm_source=bducopyrightlink&utm_medium=dswb&utm_campaign=bdu). This notebook and its source code are released under the terms of the [MIT License](https://bigdatauniversity.com/mit-license/).